File: DSA/SparshJain/Day_9_Q2.cpp
#include <iostream>
#include <stack>
#include <string>

std::string isBalanced(const std::string& s) {
    std::stack<char> stack;
    for (char ch : s) {
        if (ch == '(' || ch == '{' || ch == '[') {
            stack.push(ch);
        } else {
            if (stack.empty()) {
                return "NO";
            }
            char top = stack.top();
            if ((ch == ')' && top == '(') ||
                (ch == '}' && top == '{') ||
                (ch == ']' && top == '[')) {
                stack.pop();
            } else {
                return "NO";
            }
        }
    }
    return stack.empty() ? "YES" : "NO";
}

int main() {
    std::string s = "{[()]}";
    std::cout << isBalanced(s) << std::endl; // Output: YES

    s = "{[(])}";
    std::cout << isBalanced(s) << std::endl; // Output: NO

    return 0;
}

File: DSA/Shruti-Pandey/Day_9_q2.java
import java.util.Stack;
import java.util.Scanner;

public class BalancedBrackets {
    public static String isBalanced(String s) {
        Stack<Character> stack = new Stack<>();
        
        for (char c : s.toCharArray()) {
            if (c == '(' || c == '{' || c == '[') {
                stack.push(c);
            } else if (c == ')' || c == '}' || c == ']') {
                if (stack.isEmpty()) {
                    return "NO";
                }
                char top = stack.pop();
                if ((c == ')' && top != '(') ||
                    (c == '}' && top != '{') ||
                    (c == ']' && top != '[')) {
                    return "NO";
                }
            }
        }
        
        return stack.isEmpty() ? "YES" : "NO";
    }

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        
        System.out.println("Enter a string of brackets (using (), {}, and []):");
        String input = scanner.nextLine();
        
        String result = isBalanced(input);
        System.out.println(result);
        
        scanner.close();
    }
}
